### PR TITLE
improve: add Python 3.12 support (#3033)

### LINKED ISSUE #2959: chore/Allow python 3.12
Currently unstructured pins the [python version to >=3.9.0,<3.12](https://github.com/Unstructured-IO/unstructured/blob/main/setup.py#L87C22-L87C35). It would be great if 3.12 would be allowed too. Seems like there were two attempts to enable it before:
* https://github.com/Unstructured-IO/unstructured/pull/1703
* https://github.com/Unstructured-IO/unstructured/pull/1902

But the logs from the CI/CD are gone, so I am not sure why they failed.

### PR DESCRIPTION
**Pull request by MthwRobinson**
### Summary

Closes #2959. Updates the dependency and CI to add support for Python 3.12. 

The MongoDB ingest tests were disabled due to jobs like [this one](https://github.com/Unstructured-IO/unstructured/actions/runs/9133383127/job/25116767333) failing due to issues with the `bson` package. `bson` is a dependency for the AstraDB connector, but `pymongo` does not work when `bson` is installed from `pip`. This issue is documented by MongoDB [here](https://pymongo.readthedocs.io/en/stable/installation.html). Spun off #3049 to resolve this. Issue seems unrelated to Python 3.12, though unsure why this didn't surface previously.

Disables the `argilla` tests because `argilla` does not yet support Python 3.12. We can add the `argilla` tests back in once the PR references below is merged. You can still use the `stage_for_argilla` function if you're on `python<3.12` and you install `argilla` yourself.
- https://github.com/argilla-io/argilla/pull/4837

### CODE REVIEW

**Comment by MthwRobinson** [comment]
Looks like there's a package version issue that needs to be run down for the GCS connector:
```python
Exception ignored in: <function AbstractBufferedFile.__del__ at 0x7fe177b8f3a0>
Traceback (most recent call last):
  File "/home/runner/work/unstructured/unstructured/.venv/lib/python3.9/site-packages/fsspec/spec.py", line 1971, in __del__
    self.close()
  File "/home/runner/work/unstructured/unstructured/.venv/lib/python3.9/site-packages/fsspec/spec.py", line 1949, in close
    self.flush(force=True)
  File "/home/runner/work/unstructured/unstructured/.venv/lib/python3.9/site-packages/fsspec/spec.py", line 1808, in flush
    self._initiate_upload()
  File "/home/runner/work/unstructured/unstructured/.venv/lib/python3.9/site-packages/gcsfs/core.py", line 1509, in _initiate_upload
    self.content_type,
AttributeError: 'GCSFile' object has no attribute 'content_type'
```

File: requirements/ingest/gcs.txt [hunk 1]
```diff
@@ -35,37 +35,19 @@ frozenlist==1.4.1
...
     #   aiohttp
     #   aiosignal
-fsspec==2024.3.1
+fsspec==2024.5.0
     # via
+    #   -c ./ingest/../deps/constraints.txt
     #   -r ./ingest/gcs.in
     #   gcsfs
-gcsfs==2024.3.1
+gcsfs==0.8.0
```
Comment (MthwRobinson) [inline requirements/ingest/gcs.txt:43]: Not sure where this `0.8.0` is coming from, don't see that in the [version release history](https://pypi.org/project/gcsfs/#history) for that package PyP.

**Comment by MthwRobinson** [comment]
Looks like the Mongo failures are related to [this](https://pymongo.readthedocs.io/en/stable/installation.html). The Mongo client ships with its own version of `bson`, and the `bson` installed with `astra` causes a failure.

**Comment by MthwRobinson** [comment]
Captured the issue related to `bson` in #3049.

**Review by scanny — APPROVED** [review]
LGTM :)

**Pull request merged**